# Advent of Code 2025 Day 3: Lobby

## Import libraries

## Part 1

### Description Part 1

You descend a short staircase, enter the surprisingly vast lobby, and are quickly cleared by the security checkpoint. When you get to the main elevators, however, you discover that each one has a red light above it: they're all offline.

"Sorry about that," an Elf apologizes as she tinkers with a nearby control panel. "Some kind of electrical surge seems to have fried them. I'll try to get them online soon."

You explain your need to get further underground. "Well, you could at least take the escalator down to the printing department, not that you'd get much further than that without the elevators working. That is, you could if the escalator weren't also offline."

"But, don't worry! It's not fried; it just needs power. Maybe you can get it running while I keep working on the elevators."

There are batteries nearby that can supply emergency power to the escalator for just such an occasion. The batteries are each labeled with their joltage rating, a value from 1 to 9. You make a note of their joltage ratings (your puzzle input). For example:

987654321111111
811111111111119
234234234234278
818181911112111
The batteries are arranged into banks; each line of digits in your input corresponds to a single bank of batteries. Within each bank, you need to turn on exactly two batteries; the joltage that the bank produces is equal to the number formed by the digits on the batteries you've turned on. For example, if you have a bank like 12345 and you turn on batteries 2 and 4, the bank would produce 24 jolts. (You cannot rearrange batteries.)

You'll need to find the largest possible joltage each bank can produce. In the above example:

In 987654321111111, you can make the largest joltage possible, 98, by turning on the first two batteries.
In 811111111111119, you can make the largest joltage possible by turning on the batteries labeled 8 and 9, producing 89 jolts.
In 234234234234278, you can make 78 by turning on the last two batteries (marked 7 and 8).
In 818181911112111, the largest joltage you can produce is 92.
The total output joltage is the sum of the maximum joltage from each bank, so in this example, the total output joltage is 98 + 89 + 78 + 92 = 357.

There are many batteries in front of you. Find the maximum joltage possible from each bank; what is the total output joltage?

### Import data

In [1]:
with open('day_3_input.txt') as input:
    banks = [line.strip() for line in input.read().split('\n') if line.strip()]

In [2]:
banks[:10]

['4123535244222342322334342233754335452333242522124322242423331132232242422443224231234323332243364522',
 '2633322623185223292633232342336241353472323432337144252444247533312232433423423228632337331231223633',
 '5475665767681776387676746666636737456375666764936466526665745975592886647576766377678785567245677555',
 '7632362525633465259457646556334542742966514383538566582522246342353467256715675721556536637573534256',
 '4433233445334333433332243332243333323333244341233329233322351213324333213433242123334332332622363223',
 '4243433413423442221343334425453334333433434233464324443433323353143434343233444343243332444351232343',
 '2244326526434332132423332342213512442441234234442222233422443424222333232422426434522221223322522245',
 '7552753585246737877374567137454634234388987727862338637768474745434466841859283582327475573893743259',
 '2241222222231222232231212533323232334312212432322332332233222323327213224222126322321413722323911223',
 '35324542263353743434243545543334234433413635354255433

In [3]:
# Create test set
test_list = ['987654321111111', '811111111111119', '234234234234278', '818181911112111']

test_dic = {'987654321111111': 98, '811111111111119': 89, '234234234234278': 78, '818181911112111': 92}

print(test_list, test_dic)

['987654321111111', '811111111111119', '234234234234278', '818181911112111'] {'987654321111111': 98, '811111111111119': 89, '234234234234278': 78, '818181911112111': 92}


### Write function to find highest leftmost integer then highest integer to right of it

In [4]:
# Find the highest digit in string (as long as it's not the last digit)
# Iterate through string with while, split string at highest digit & keep right side
# Find the highest digit in new string
# Add it to array of bank joltages
# Return the sum of the array

def bank_joltage_sum(input):

    # Create empty list to store the joltage for each bank
    joltage_arr = []
    # Iterate through the list of banks
    for bank in input:
        # Create an object missing the last digit to find highest digit in string
        bank_no_last = bank[:-1]
        # Find the highest digit in the string
        high_battery = max(bank_no_last)
        # Split the string at the highest digit
        right_batteries = bank.split(high_battery, 1)[1]
        # Find the highest digit in the split string
        high_battery_right = max(right_batteries)
        # Combine the two digits as strings (i.e. '1' + '3' = '13')
        joltage = int(str(high_battery) + str(high_battery_right))
        # Add the joltage to the array of joltages
        joltage_arr += [joltage]
    # Sum the joltage array for total
    joltage_sum = sum(joltage_arr)
            
    return joltage_sum

In [5]:
bank_joltage_sum(banks)

17158

## Solution: The total output joltage is 17,158

## Part 2

### Description Part 2

The escalator doesn't move. The Elf explains that it probably needs more joltage to overcome the static friction of the system and hits the big red "joltage limit safety override" button. You lose count of the number of times she needs to confirm "yes, I'm sure" and decorate the lobby a bit while you wait.

Now, you need to make the largest joltage by turning on exactly twelve batteries within each bank.

The joltage output for the bank is still the number formed by the digits of the batteries you've turned on; the only difference is that now there will be 12 digits in each bank's joltage output instead of two.

Consider again the example from before:

987654321111111
811111111111119
234234234234278
818181911112111
Now, the joltages are much larger:

In 987654321111111, the largest joltage can be found by turning on everything except some 1s at the end to produce 987654321111.
In the digit sequence 811111111111119, the largest joltage can be found by turning on everything except some 1s, producing 811111111119.
In 234234234234278, the largest joltage can be found by turning on everything except a 2 battery, a 3 battery, and another 2 battery near the start to produce 434234234278.
In 818181911112111, the joltage 888911112111 is produced by turning on everything except some 1s near the front.
The total output joltage is now much larger: 987654321111 + 811111111119 + 434234234278 + 888911112111 = 3121910778619.

What is the new total output joltage?

### Write function 

In [6]:
# Find the highest digit in string (as long as it's not in last 11 digits)
# Store in an object (called joltage) as string
# Split the whole bank string at highest digit
# Find max digit in that, repeat until len(split_string) = 12 - len(joltage)

def twelve_digit_joltage(input):

    # Create empty list to store the joltage for each bank
    joltage_arr = []
    # Iterate through the list of banks
    for bank in input:
        # Create an object missing the last digit to find highest digit in string
        bank_no_last = bank[:-11]
        # Find the highest digit in the string
        high_battery = max(bank_no_last)
        # Add the highest digit to the joltage object
        joltage = str(high_battery)
        # Split the string at the highest digit
        right_batteries = bank.split(high_battery, 1)[1]
        # Set the number of missing batteries from the string
        missing_batteries = -10
        # Continue to iterate through the bank until the bank is too short to split (need 12 digits total)
        while len(right_batteries) > (12 - len(joltage)) and len(joltage) < 12:
            # Make a special case for the last digit in the joltage
            if len(joltage) == 11:
                # Find the highest digit in the whole string (no characters needed to right)
                high_battery = max(right_batteries)
                # Combine the two digits as strings (i.e. '1' + '3' = '13')
                joltage += str(high_battery)
            else:
                # Limit the search for highest digit to the # missing digits from joltage
                bank_no_last = right_batteries[:(missing_batteries)]
                # Find the highest digit in the string
                high_battery = max(bank_no_last)
                # Combine the two digits as strings (i.e. '1' + '3' = '13')
                joltage += str(high_battery)
                # Split the string at the highest digit
                right_batteries = right_batteries.split(high_battery, 1)[1]
                # Go to the next battery
                missing_batteries += 1
        # If the string is too short to split, add the whole thing to the joltage
        if len(joltage) < 12:
            joltage += str(right_batteries)
        # Make joltage an integer to add to array (need to sum later)
        joltage = int(joltage)
        # Add the joltage to the array of joltages
        joltage_arr += [joltage]
    # Sum the joltage array for total
    joltage_sum = sum(joltage_arr)

    return(joltage_sum)

In [7]:
twelve_digit_joltage(banks)

170449335646486

## Solution: The total output joltage is 170,449,335,646,486